src/include/catalog/catalog_accessor.h

```diff
@@ -0,0 +1,280 @@
+   * @return the storage object corresponding to the passed OID
+   */
+  common::ManagedPointer<storage::SqlTable> GetTable(table_oid_t table);
+
+  /**
+   * Apply a new schema to the given table.  The changes should modify the latest
+   * schema as provided by the catalog.  There is no guarantee that the OIDs for
+   * modified columns will be stable across a schema change.
+   * @param table OID of the modified table
+   * @param new_schema object describing the table after modification
+   * @return true if the operation succeeded, false otherwise
+   * @warning The catalog accessor assumes it takes ownership of the schema object
+   * that is passed.  As such, there is no guarantee that the pointer is still
+   * valid when this function returns.  If the caller needs to reference the
+   * schema object after this call, they should use the GetSchema function to
+   * obtain the authoritative schema for this table.
+   */
+  bool UpdateSchema(table_oid_t table, Schema *new_schema);
+
+  /**
+   * Get the visible schema describing the table.
+   * @param table corresponding to the requested schema
+   * @return the visible schema object for the identified table
+   */
+  const Schema &GetSchema(table_oid_t table);
+
+  /**
+   * A list of all constraints on this table
+   * @param table being queried
+   * @return vector of OIDs for all of the constraints that apply to this table
+   */
+  std::vector<constraint_oid_t> GetConstraints(table_oid_t table);
+
+  /**
+   * A list of all indexes on the given table
+   * @param table being queried
+   * @return vector of OIDs for all of the indexes on this table
+   */
+  std::vector<index_oid_t> GetIndexes(table_oid_t table);
+
+  /**
+   * Given an index name, resolve it to the corresponding OID
+   * @param name of the index
+   * @return OID of the index, INVALID_INDEX_OID if the index was not found
+   */
+  index_oid_t GetIndexOid(const std::string &name);
+
+  /**
+   * Given an index name and the owning namespace, resolve it to the corresponding OID
+   * @param ns in which to search for the index
+   * @param name of the index
+   * @return OID of the index, INVALID_INDEX_OID if the index was not found
+   */
+  index_oid_t GetIndexOid(namespace_oid_t ns, const std::string &name);
+
+  /**
+   * Given a table, find all indexes for data in that table
+   * @param table OID being queried
+   * @return vector of index OIDs that reference the queried table
+   */
+  std::vector<index_oid_t> GetIndexOids(table_oid_t table);
+
+  /**
+   * Given the index name and its specification, add it to the catalog
+   * @param ns is the namespace in which the index will exist
+   * @param table on which this index exists
+   * @param name of the index
+   * @param schema describing the new index
+   * @return OID for the index, INVALID_INDEX_OID if the operation failed
+   */
+  index_oid_t CreateIndex(namespace_oid_t ns, table_oid_t table, const std::string &name, const IndexSchema &schema);
```

Comment:
Aren't the table's namespace and the index's namespace the same? If so, is the first parameter needed?

Reply:
There is no requirement that an index and the table it indexes have to be in the same namespace, so we need to have it as a parameter here.